New tests and updated readme

beccaelenzil opened the pull request:
The tests have been updated to mostly use data in `tests_constants.py.`
- The user_data for each wave builds on the user_data from previous waves so that the data structure is consistent throughout.
- The full test suite is provided here. Once we are confident this is test suite is complete, I will remove some of the steps from the tests (Waves 02 and later) to scaffold students learning to write tests.

The README has been updated to emphasize TDD
- As part of the workflow, students need to check if a test is incomplete, and if it is, complete it
- I made the decision to remove the extra section on submission/workflow because it seemed to repeat the workflow above, but simplify it in a way that leaves out steps.
- A `play_tester.py` file is provided and there are suggestions on how to use it

This fork/branch has a complete solution with this test suite: https://github.com/beccaelenzil/viewing-party/tree/be/new-tests

goeunpark commented on tests/test_wave_03.py:
```diff
@@ -1,212 +1,77 @@
-                    }
-                ]
-            }
-        ]
-    }
+    amandas_data = USER_DATA_3
```
something something pytest fixtures instead of using mutable global variables? (sry, still a a python n00b ;__; )

goeunpark replied:
update: this is likely happening because my implementation of `get_unique_watched` mutates the `user_data` 😬

beccaelenzil replied:
per our conversation, we don't want students to mutate `user_data`, but it's also a problem that our unit tests don't start fresh data each time. I will think more on how to handle this. Introducing fixtures now feels too early, so it may mean copying and pasting the data into each test.

beccaelenzil replied:
The approach of pulling the data out into a `test_constants.py` file came from writing integration tests. We decided to scrap the integration tests because they were really unit tests stuck together, but I think this same approach doesn't work well for unit tests.

goeunpark commented on README.md:
````diff
@@ -141,19 +156,26 @@ $ pytest tests/test_file_name.py
+```
 
-At submission time, no matter where you are, submit the project via Learn.
+There is some starter code provided in `play_tester.py`. This code prints the test data that is used for many of the tests. Looking closely at this data can help us think critically about the expected output for given input for each function. Then, calling each function with this data allows us to observe the **actual** output for given input. 
````
this is so helpful, esp since we cleaned up our test files and can no longer just check the shape of the input by glancing at the test

kaidamasaki commented on tests/test_wave_01.py:
```diff
@@ -154,33 +142,17 @@ def test_moves_movie_from_watchlist_to_watched():
 
     # Act
-    updated_data = watch_movie(janes_data, "Title A")
+    updated_data = watch_movie(janes_data, "Non-Existent Movie Title")
```
I really like this clarification in the test.

kaidamasaki reviewed (commented):
Looks pretty good.  Just a few suggestions, mostly around not using `is` and making deep copies.

beccaelenzil commented on README.md:
```diff
@@ -85,54 +87,67 @@ $ source venv/bin/activate
 2. Find the test file that contains the test you want to run. Ensure that the test(s) you want to run isn't skipped.
 
-   - Check the `tests` folder, and find the test file you want to run
+    - Check the `tests` folder, and find the test file you want to run
```
Actually no! It reads a little bit more clearly with the double indent

beccaelenzil commented on README.md:
```diff
@@ -85,54 +87,67 @@ $ source venv/bin/activate
    - In that test file, read through each test case
-   - Remove all lines that contain `@pytest.mark.skip()`
+      - If it is incomplete, complete the test. 
+        - *Is this a nominal or edge case?*
```
Which lesson? Intro to Automated Tests or Intro to Pytest

beccaelenzil replied:
Can't link to lessons from repos since students are in different classes.

beccaelenzil commented:
Based on feedback, the following changes were made:
* Replaced is with ==
* Implemented functions to provide a deep copy of USER_DATA for each test to mimic fixtures
* Updated Readme to describe test_constants
* Update Readme to add note about not modifying the data

beccaelenzil commented:
Here is a separate PR removing act and/or assert for 4 tests (Waves 01, 03, and 05): https://github.com/beccaelenzil/viewing-party/pull/2/files

goeunpark commented on README.md:
````diff
@@ -141,19 +158,32 @@ $ pytest tests/test_file_name.py
-At submission time, no matter where you are, submit the project via Learn.
+To run this file, use:
+
+```bash
+python3 play_tester.py
+```
````
This also has to be run in venv, might be worth adding `# Must be in activated virtual environment` NBD if it's overkill though

beccaelenzil commented:
Merged with this PR: https://github.com/AdaGold/viewing-party/pull/18
